generators/beamAndVertexGenerator.cc

```diff
@@ -118,80 +201,201 @@ void beamAndVertexGenerator::randomizeBeamfileStartingPosition() {
-			               z);
-			_beam.SetXYZM(randomGen->Gaus(_beamMomentumX.first, _sigmaScalingFactor * _beamMomentumX.second),
-			              randomGen->Gaus(_beamMomentumY.first, _sigmaScalingFactor * _beamMomentumY.second),
-			              randomGen->Gaus(_beamMomentumZ.first, _sigmaScalingFactor * _beamMomentumZ.second),
-						  beam.particle.mass());
-		} else {
-			_vertex.SetXYZ(projectedVertexX, projectedVertexY, z);
-			_beam.SetXYZM(_beamMomentumX.first,
-			              _beamMomentumY.first,
-			              _beamMomentumZ.first,
-			              beam.particle.mass());
-		}
+		_vertex.SetXYZ(projectedVertexX, projectedVertexY, z);
+		_beam.SetXYZM(_beamMomentumX,
+						_beamMomentumY,
+						_beamMomentumZ,
+						beam.particle.mass());
 	}
 	return true;
 }
 
 
-double beamAndVertexGenerator::getVertexZ(const Target& target) const {
+bool
+beamAndVertexGenerator::eventFromMomentaCorrectMomentumResolution(const Target& target, const Beam& beam) {
+	TRandom3* randomGen = randomNumberGenerator::instance()->getGenerator();
+	if(not loadNextEventFromBeamfile()) return false;
+
+	const double z = getVertexZ(target); // has to be done AFTER loading the next event
+
+	const double dx = _beamMomentumX / _beamMomentumZ;
+	const double dy = _beamMomentumY / _beamMomentumZ;
+	const double projectedVertexX = _vertexX + dx * (z-_vertexZ);
+	const double projectedVertexY = _vertexY + dy * (z-_vertexZ);
+
+	_beamMomentum = sqrt(_beamMomentumX*_beamMomentumX + _beamMomentumY*_beamMomentumY + _beamMomentumZ*_beamMomentumZ);
+	const double truthMomentumPDFSigma = sqrt(1.0/(1.0/_momentumResolution/_momentumResolution+ 1.0/_momentumPDFSigma/_momentumPDFSigma));
+	const double truthMomentumPDFMean = truthMomentumPDFSigma*truthMomentumPDFSigma*( 1./_momentumResolution/_momentumResolution*_beamMomentum
+																					+ 1./_momentumPDFSigma/_momentumPDFSigma*_momentumPDFMean );
+	const double truthMomentum = randomGen->Gaus(truthMomentumPDFMean, truthMomentumPDFSigma);
+
+	_vertex.SetXYZ(projectedVertexX, projectedVertexY, z);
+	const double norm = sqrt(dx*dx + dy*dy + 1);
+	_beam.SetXYZM(truthMomentum * dx / norm,
+	              truthMomentum * dy / norm,
+	              truthMomentum * 1 / norm,
+	              beam.particle.mass());
+	return true;
+}
+
+
+bool
+beamAndVertexGenerator::eventFromInclinations(const Target& target, const Beam& beam) {
 	TRandom3* randomGen = randomNumberGenerator::instance()->getGenerator();
-	double z;
-	do {
-		z = randomGen->Exp(target.interactionLength);
-	} while(z > target.length);
-	z = (target.position.Z() - target.length * 0.5) + z;
-	return z;
+	if(not loadNextEventFromBeamfile()) return false;
+
+	const double z = getVertexZ(target); // has to be done AFTER loading the next event
+
+	TVectorD mean(4);
+	mean[0] = _vertexX;
+	mean[1] = _vertexY;
+	mean[2] = _beamdXdZ;
+	mean[3] = _beamdYdZ;
+	TVectorD sample = multivariantGauss(mean, *_vertexCovarianceMatrix, _sigmaScalingFactor);
```

Comment:
in L273 you checked _sigmasPresent and _sigmaScalingFactor != 0. This is not needed here?

Reply:
_sigmasPresent does not need to be checked, because for beam files with inclinations, the sigma branches are required!
If _sigmaScalingFactor = 0, the Gauss is simply a delta function. At the moment, ROOT can handle this, but I add the check in order to avoid problems in the future!